### ISSUE TITLE
Error: Kerberos library is not installed

### ISSUE THREAD

**Issue by EdikDolynskyi** [christkv/mongodb-core#127]
Hello Christian.

I am currently working on a project which requires me to implement connection to MongoDB via different authentication services. I have decided to use "mongodb-core" and "kerberos" modules to integrate this functionality.

I am very thankful to you for these modules, they were of much help to me while developing handling of different connections.

---

However, I experience following error trying to setup Kerberos connection to MongoDB:

``` Error: Kerberos library is not installed ```

I tracked this error to be happening in this place (mongodb-core\lib\auth\gsapi.js):

```
try {
  	Kerberos = require_optional('kerberos').Kerberos;
  	// Authentication process for Mongo
  	MongoAuthProcess = require_optional('kerberos').processes.MongoAuthProcess
} 
```

Here are the connection params I use:

**Connection URI:** 
``` mongodb://steve%252Fadmin%2540EXAMPLE.COM:123@server.example.com:27017/?slaveOk=true&gssapiServiceName=mongodb&authMechanism=GSSAPI ```

**Connection Options:**
```
{ 
     uri_decode_auth: true,
     db: { readPreference: 'secondaryPreferred' },
     replSet: { connectWithNoPrimary: true }
 }
```

I am using:
*"mongodb": "2.2.5",*
*"mongodb-core": "2.0.7"*
*"kerberos": "0.0.21"*

As for the Kerberos infrastructure, I have it all set up and I am able to connect to mongod through mongo shell and $external authentication.

Could you please point to what I am possibly doing wrong?

Thanks,
Edik.

**Comment by christkv** [comment]
That means the kerberos library is not available. You need to add the kerberos library to your package.json and install it. If you see no compilation issues then you won't have any problems.

However if you are on windows you will need to set up a windows 64 bit sdk and compiler tool chain.

**Comment by EdikDolynskyi** [comment]
Thanks for the tip. I have installed kerberos library with npm into my project, but it didn't help. Here's installation log I have (I am trying it on Ubuntu):

```
> kerberos@0.0.21 install /home/eduard/Develop/binary-studio/node_modules/kerberos
> (node-gyp rebuild) || (exit 0)

gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
make: Entering directory `/home/eduard/Develop/binary-studio/node_modules/kerberos/build'
  CXX(target) Release/obj.target/kerberos/lib/kerberos.o
  CXX(target) Release/obj.target/kerberos/lib/worker.o
  CC(target) Release/obj.target/kerberos/lib/kerberosgss.o
../lib/kerberosgss.c:36:0: warning: ignoring #pragma clang diagnostic [-Wunknown-pragmas]
 #pragma clang diagnostic push
 ^
../lib/kerberosgss.c:37:0: warning: ignoring #pragma clang diagnostic [-Wunknown-pragmas]
 #pragma clang diagnostic ignored "-Wdeprecated-declarations"
 ^
../lib/kerberosgss.c: In function ‘authenticate_gss_client_wrap’:
../lib/kerberosgss.c:446:19: warning: variable ‘server_conf_flags’ set but not used [-Wunused-but-set-variable]
   char buf[4096], server_conf_flags;
                   ^
../lib/kerberosgss.c: At top level:
../lib/kerberosgss.c:1236:0: warning: ignoring #pragma clang diagnostic [-Wunknown-pragmas]
 #pragma clang diagnostic pop
 ^
  CC(target) Release/obj.target/kerberos/lib/base64.o
  CXX(target) Release/obj.target/kerberos/lib/kerberos_context.o
  SOLINK_MODULE(target) Release/obj.target/kerberos.node
  COPY Release/kerberos.node
make: Leaving directory `/home/eduard/Develop/binary-studio/node_modules/kerberos/build'
Hackolade@1.0.6 /home/eduard/Develop/binary-studio
├── UNMET PEER DEPENDENCY exports-loader@>=0.6.2
├── UNMET PEER DEPENDENCY imports-loader@>=0.6.3
├─┬ kerberos@0.0.21 
│ └── nan@2.3.5 
└── UNMET PEER DEPENDENCY less-loader@>=0.6.2
```

After this installation the error persists.

**Comment by EdikDolynskyi** [comment]
Hi, again! Maybe  you have any idea how to resolve this issue?

**Comment by christkv** [comment]
can you post the

ls -la node_modules

**Comment by christkv** [comment]
I just set up a test project and cannot reproduce this issue I'm adding the file I'm using (you will have to change your own credentials obviously and supply your own ops since this is against an internal test system).

[Archive.zip](https://github.com/christkv/mongodb-core/files/436483/Archive.zip)

**Comment by EdikDolynskyi** [comment]
Thanks a lot for an answer, it was helpful for me to find a problem. 
Actually, the Kerberos was installed on my PC and was working fine.

It was an error related to different NodeJS versions used by npm modules. That’s why it couldn’t find the right way to Kerberos inside modules. The electron-rebuild resolved this issue.

Many thanks again.

**Comment by christkv** [comment]
no problem

**Comment by polkhovsky** [comment]
Somebody may find this helpful:
I recently experienced the same issue. Mongodb-core hides the error when the kerberos module's version is higher than the required dependency version. In my case 0.0.24 was installed but:
`Error: optional dependency [kerberos] found but version [0.0.24] did not satisfy constraint [^0.0.23]`
Once I downgraded the version, everything worked fine. [More details on Stack Overflow.](https://stackoverflow.com/questions/50844414/kerberos-library-is-not-installed)
